File: leetcode/Algorithms/494.TargetSum/Solution.java
public class Solution {
    private int result = 0;
    private int target;
    private int[] tmp;

    public int findTargetSumWays(int[] nums, int S) {
        target = S;
        tmp = nums;
        travel(0, 0);
        return result;
    }

    private void travel(int t, int index) {
        if (index == tmp.length - 1) {
            if (t + tmp[index] == target) result++;
            if (t - tmp[index] == target) result++;
        } else if (index < tmp.length - 1) {
            travel(t + tmp[index], index + 1);
            travel(t - tmp[index], index + 1);
        }
    }
}

File: leetcode/Algorithms/494.TargetSum/Solution.py
class Solution(object):
    def findTargetSumWays(self, nums, S):
        """
        :type nums: List[int]
        :type S: int
        :rtype: int
        """
        all_sum = sum(nums)
        if S < -all_sum or S > all_sum:
            return 0

        dp = [0] * (all_sum * 2 + 1)
        dp[all_sum] = 1
        for num in nums:
            tmp = [0] * (all_sum * 2 + 1)
            for i in range(len(dp)):
                if dp[i] != 0:
                    tmp[i + num] += dp[i]
                    tmp[i - num] += dp[i]
            dp = tmp

        return dp[S + all_sum]

